### PR TITLE
rustdoc: Don't record associated items from trait impls in the search-index

### LINKED ISSUE #31808: rustdoc: implementing trait with associated const for primitive causes panic (regression)
Nightly rustdoc seems to panic when implementing a trait with an associated const for a primitive type.

Reduced test case:

```rust
#![feature(associated_consts)]

pub trait Foo {
    const BAR: usize;
}

impl Foo for () {
    const BAR: usize = 0;
}
```

Message:
```thread '<unnamed>' panicked at 'called `Option::unwrap()` on a `None` value', ../src/libcore\option.rs:330```

## Meta

```
>rustdoc -vV
rustdoc 1.8.0-nightly (57c357d89 2016-02-16)
binary: rustdoc
commit-hash: 57c357d89183df173b0e42a0f745d9cfdb67fb1a
commit-date: 2016-02-16
host: x86_64-pc-windows-msvc
release: 1.8.0-nightly
```

Backtrace:
```
   0:     0x7ffc6b5cd947 - panicking::Location<'a>::line::hd292ced6adc3cf6b7fz
   1:     0x7ffc6b5cd146 - panicking::take_handler::h49f661ca9b556367Wdz
   2:     0x7ffc6b59410f - sys_common::unwind::begin_unwind_inner::h72177c2eb1fe6c2eLct
   3:     0x7ffc6b594b9d - sys_common::unwind::begin_unwind_fmt::h9befc5f9bdb15c19Rbt
   4:     0x7ffc6b5c2f82 - rust_begin_unwind
   5:     0x7ffc6b5daac8 - panicking::panic_fmt::h8430b6988b78563f6YL
   6:     0x7ffc6b5db3c0 - panicking::panic::h331ae34734277dc3DXL
   7:     0x7ffc669c128e - html::render::Cache.DocFolder::fold_item::h467e95350f348073Q5o
   8:     0x7ffc6684b7d1 - html::render::run::h9521371f3809b50bX8n
   9:     0x7ffc6682626a - main_args::h6d07868c1334373bU3u
  10:     0x7ffc66822751 - main::hf5a3402ea1356a63GZu
  11:     0x7ffc6b5c2eb8 - sys_common::thread_local::Key.Drop::drop::h122da565f997e58984s
  12:     0x7ffc6b5bc7f2 - sys_common::unwind::inner_try::h6f3bf9f042fec881N9s
  13:     0x7ffc66822cdf - main::hf5a3402ea1356a63GZu
  14:     0x7ffc6b5cb62b - sys::process::Command::cwd::h2628fdc5c60e66846Sx
  15:     0x7ffc860f8101 - BaseThreadInitThunk
```

### PR DESCRIPTION
**Pull request by mitaa**
This effectively only records associated items from either inherent impls or trait definitions in the search-index.

fixes #31808 

r? @alexcrichton

### CODE REVIEW

File: src/librustdoc/html/render.rs [hunk 1]
```diff
@@ -1118,9 +1121,11 @@ impl DocFolder for Cache {
         let parent_pushed = match item.inner {
             clean::TraitItem(..) | clean::EnumItem(..) | clean::StructItem(..) => {
                 self.parent_stack.push(item.def_id);
+                self.parent_is_trait_impl = false;
                 true
             }
             clean::ImplItem(ref i) => {
+                self.parent_is_trait_impl = i.trait_.is_some();
```
Comment (alexcrichton) [inline src/librustdoc/html/render.rs]: I think you can have an impl-in-an-impl, so this flag may need to be preserved and restored across the recursive calls?

**Comment by mitaa** [comment]
I pushed an update restoring the flag across recursive calls.

File: src/librustdoc/html/render.rs [hunk 2]
```diff
@@ -1140,6 +1145,7 @@ impl DocFolder for Cache {
             }
             _ => false
         };
+        let parent_is_trait_impl = self.parent_is_trait_impl;
```
Comment (alexcrichton) [inline src/librustdoc/html/render.rs]: Shouldn't this be saved before the `match` above which modifies it?

**Comment by mitaa** [comment]
> Shouldn't this be saved before the match above which modifies it?

Yeah.. sorry, about that.

**Pull request merged**